### PR TITLE
[SPARK-42307][SQL] Assign name for error _LEGACY_ERROR_TEMP_2232

### PR DESCRIPTION
**Pull request by junyuc25**
### What changes were proposed in this pull request?

In this PR, I propose to replace the legacy error name `_LEGACY_ERROR_TEMP_2232` with `ROW_VALUE_IS_NULL `, and add a test case for it.

### Why are the changes needed?

Proper name improves user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Ran all the tests in the suite:
```
build/sbt "testOnly *org.apache.spark.sql.RowSuite"
```

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by junyuc25** [comment]
Hi @MaxGekk , I propose a PR to update the name for legacy errors. Do you mind taking a look? Thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala [hunk 1]
```diff
@@ -123,4 +123,17 @@ class RowSuite extends SparkFunSuite with SharedSparkSession {
       parameters = Map("methodName" -> "fieldIndex", "className" -> "Row", "fieldName" -> "`foo`")
     )
   }
+
+  test("SPARK-42307 - Trying to get the value from a null column should result in error") {
```
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala:127]: nit:
```suggestion
  test("SPARK-42307: get the value from a null column should result in error") {
```

**Review by allisonwang-db — APPROVED** [review]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala (hunk 1 above)
Comment (junyuc25) [reply sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala]: Thanks for the comment @allisonwang-db. I updated the test description.

Comment (junyuc25) [reply sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala]: My bad I clicked the revoke review by mistake...

Comment (junyuc25) [reply sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala]: cc @allisonwang-db @cloud-fan @MaxGekk. Not sure if you got a chance to review this small PR? Thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala [hunk 2]
```diff
@@ -123,4 +123,17 @@ class RowSuite extends SparkFunSuite with SharedSparkSession {
       parameters = Map("methodName" -> "fieldIndex", "className" -> "Row", "fieldName" -> "`foo`")
     )
   }
+
+  test("SPARK-42307 - get the value from a null column should result in error") {
```
Comment (allisonwang-db) [inline sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala:127]:
```suggestion
  test("SPARK-42307: get a value from a null column should result in error") {
```

Comment (junyuc25) [reply sql/core/src/test/scala/org/apache/spark/sql/RowSuite.scala]: Thanks @allisonwang-db! Updated.

File: common/utils/src/main/resources/error/error-conditions.json [hunk 3]
```diff
@@ -572,6 +572,12 @@
     ],
     "sqlState" : "42703"
   },
+  "COLUMN_VALUE_IS_NULL" : {
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:575]: Let's name in more precise way:
```suggestion
  "ROW_VALUE_IS_NULL" : {
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 4]
```diff
@@ -572,6 +572,12 @@
...
+    "message" : [
+      "Value at index <index> is null."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:577]:
```suggestion
Found NULL in a row at the index <index>, expected a non-NULL value.
```

File: common/utils/src/main/resources/error/error-conditions.json [hunk 5]
```diff
@@ -572,6 +572,12 @@
...
+    ],
+    "sqlState" : "22000"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-conditions.json:579]: How about this number "22023"

**Review by MaxGekk — COMMENTED** [review]
Is it correct? If not, please, fix it.
https://github.com/apache/spark/blob/731a104df30751ca0c333e621129564f34b81994/sql/api/src/main/scala/org/apache/spark/sql/Row.scala#L526

File: common/utils/src/main/resources/error/error-conditions.json (hunk 5 above)
Comment (junyuc25) [reply common/utils/src/main/resources/error/error-conditions.json]: Done

File: common/utils/src/main/resources/error/error-conditions.json (hunk 3 above)
Comment (junyuc25) [reply common/utils/src/main/resources/error/error-conditions.json]: Updated.

File: common/utils/src/main/resources/error/error-conditions.json (hunk 4 above)
Comment (junyuc25) [reply common/utils/src/main/resources/error/error-conditions.json]: Done

File: sql/api/src/main/scala/org/apache/spark/sql/Row.scala [hunk 6]
```diff
@@ -219,39 +219,34 @@ trait Row extends Serializable {
    * Returns the value at position i as a primitive boolean.
    *
    * @throws ClassCastException when data type does not match.
-   * @throws NullPointerException when value is null.
```
Comment (MaxGekk) [inline sql/api/src/main/scala/org/apache/spark/sql/Row.scala:222]: Could you clarify why did you remove it instead of replacing `NullPointerException` by `SparkException`, just wonder.

**Review by MaxGekk — COMMENTED** [review]
> ... name _LEGACY_ERROR_TEMP_2232 with COLUMN_VALUE_IS_NULL

Could you update PR's description according to your changes, please.

File: sql/api/src/main/scala/org/apache/spark/sql/Row.scala (hunk 6 above)
Comment (junyuc25) [reply sql/api/src/main/scala/org/apache/spark/sql/Row.scala:222]: Oops I misunderstood your previous comment. Replaced NPE with `SparkException`.

Comment (MaxGekk) [reply sql/api/src/main/scala/org/apache/spark/sql/Row.scala:222]: And one more thing is need to update the SQL migration guide because this might impact user apps and potentially break them. Please, add an item to `sql-migration-guide.md`.

**Comment by junyuc25** [comment]
Hi @MaxGekk , looks like the current implementation is causing an error when generating docs ([job link](https://github.com/junyuc25/spark/actions/runs/10349944541/job/28645454259)), because SparkException is declared in the comment but it is not thrown from the codes.
```
[error] /__w/spark/spark/sql/api/target/java/org/apache/spark/sql/Row.java:135:1:  error: exception not thrown: org.apache.spark.SparkException
[error]    * @throws org.apache.spark.SparkException when value is null.
[error]              ^
[warn] javadoc exited with exit code 1
```
I found two options to fix this:
1. Add `@throws[SparkException]` annotation to the method. This is equivalent to adding `throws XXException` to a Java method signature (i.e. `public  boolean getBoolean (int i) throws org.apache.spark.SparkException`), and hence Java consumers of this method would need to catch this exception explicitly.
```
  /**
   * Returns the value at position i as a primitive boolean.
   *
   * @throws ClassCastException when data type does not match.
   * @throws org.apache.spark.SparkException when value is null.
   */
  @throws[SparkException]
  def getBoolean(i: Int): Boolean = getAnyValAs[Boolean](i)
```

2. Do not add `@throws org.apache.spark.SparkException` to the comment, and also remove `@throws NullPointerException` (because NPE would not be thrown from the codes).
```
  /**
   * Returns the value at position i as a primitive boolean.
   *
   * @throws ClassCastException when data type does not match.
   */
  def getBoolean(i: Int): Boolean = getAnyValAs[Boolean](i)
```
I kind of lean towards option #1 as it could provide users with better context about the exception type being thrown, although they would have to handle this exception explicitly in their codes. Let me know your thought on this or if you have other ideas. Thanks :)!

**Comment by MaxGekk** [comment]
@junyuc25 I guess a runtime exception might not force catching the exception. How about to replace `SparkException` by `SparkRuntimeException` in `valueIsNullError()`?

**Review by MaxGekk — COMMENTED** [review]
LGTM waiting for CI.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @junyuc25 and @allisonwang-db for review.

**Comment by MaxGekk** [comment]
@junyuc25 Do you have an account at the OSS JIRA? If so, leave a comment in SPARK-42307. I will assign the ticket to you.

**Comment by junyuc25** [comment]
Yes I posted a comment in the Jira ticket. Thanks for reviewing!

Sent from Gmail Mobile


Maxim Gekk ***@***.***>于2024年8月14日 周三01:18写道：
